### PR TITLE
VITIS 2383: xbutil to show usage stats for all buffer types

### PR DESCRIPTION
**Pull request by sarab96**
**See json output & txt output info in separate comment below**

Problem solved by the commit
None. Early access feature for xbutil to show usage stats of BOs

Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
None

What has been tested and how, request additional testing if necessary
Tested Locally

Documentation impact (if any)
To be done as early access

### CODE REVIEW

**Comment by sarab96** [comment]
**txt/console output:**
image

**json output:**
{
    "schema_version": {
        "schema": "JSON",
        "creation_date": "Thu Feb 17 07:07:17 2022 GMT"
    },
    "devices": [
        {
            "interface_type": "pcie",
            "device_id": "0000:b3:00.1",
            "buffer_object_stats": [
                {
                    "buffer_type": "Regular",
                    "buffer_count": "0",
                    "memory_used": "0",
                    "memory_used_unit": "KB",
                    "memory_used_bytes": "0"
                },
                {
                    "buffer_type": "UserPointer",
                    "buffer_count": "0",
                    "memory_used": "0",
                    "memory_used_unit": "KB",
                    "memory_used_bytes": "0"
                },
                {
                    "buffer_type": "P2P",
                    "buffer_count": "0",
                    "memory_used": "0",
                    "memory_used_unit": "KB",
                    "memory_used_bytes": "0"
                },
                {
                    "buffer_type": "DeviceOnly",
                    "buffer_count": "0",
                    "memory_used": "0",
                    "memory_used_unit": "KB",
                    "memory_used_bytes": "0"
                },
                {
                    "buffer_type": "Imported",
                    "buffer_count": "0",
                    "memory_used": "0",
                    "memory_used_unit": "KB",
                    "memory_used_bytes": "0"
                },
                {
                    "buffer_type": "ExecBuf",
                    "buffer_count": "0",
                    "memory_used": "0",
                    "memory_used_unit": "KB",
                    "memory_used_bytes": "0"
                },
                {
                    "buffer_type": "CMA",
                    "buffer_count": "0",
                    "memory_used": "0",
                    "memory_used_unit": "KB",
                    "memory_used_bytes": "0"
                }
            ]
        }
    ]
}

File: src/runtime_src/core/tools/common/ReportBOStats.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      //Expected fileds are BO type, size in KB and Num of BOs
+      if (bo_info.size() != 3)
+        throw std::runtime_error((boost::format("ERROR: Unexpected format in BO Stats. Line: %s") % line).str());
+
+      boost::property_tree::ptree bo_pt;
+      if (bo_info[0].length() < 5 || bo_info[1].length() < 3 || bo_info[2].length() < 4) {
+        bo_pt.put("bufer_type", "Invalid");
+        bo_pt.put("buffer_count", "n/a");
+        bo_pt.put("memory_used", "n/a");
+        bo_pt.put("memory_used_unit", "n/a");
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/ReportBOStats.cpp:96]: We don't need this anymore. Please follow the json that Stephen provided in the email

**Review by AShivangi — CHANGES_REQUESTED** [review]
It is mostly good except for a couple of fields that should be removed from the json, like:
```
{
"buffer_type": "Regular",
"buffer_count": "0",
"memory_used_bytes": "0"
},
```

File: src/runtime_src/core/tools/common/ReportBOStats.cpp (hunk 1 above)
Comment (sarab96) [reply src/runtime_src/core/tools/common/ReportBOStats.cpp:96]: These are required for txt output on console else will require division to be performed to display in other units. And also this provides mem info in two formats for users. So gives users more convenience.

File: src/runtime_src/core/tools/common/ReportBOStats.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+ReportBOStats::writeReport(const xrt_core::device* /*pDevice*/,
+                        const boost::property_tree::ptree& pt,
+                        const std::vector<std::string>& /*elementsFilter*/,
+                        std::ostream & output) const
+{
+  boost::property_tree::ptree empty_ptree;
+
+  Table2D::HeaderData bo_type = {"Buffer Type", Table2D::Justification::right};
+  Table2D::HeaderData bo_count = {"Buffer Count", Table2D::Justification::center};
+  Table2D::HeaderData mem_used = {"Memory Usage", Table2D::Justification::center};
```
[the remark this answers is not in the document]
Comment (sarab96) [reply src/runtime_src/core/tools/common/ReportBOStats.cpp]: done

File: src/runtime_src/core/tools/common/ReportBOStats.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    bool found = false;
+
+    //Skip DDR usage info by looking for line with string "BO Stats Below"
+    //Following that are BO stats. BO type, size in KB and num of BOs fields separated by blank space
+    for (const auto& line: mem_stat) {
+      if (line.empty())
+        continue;
+
+      //Skip DDR usage stats at begining by looking for beloe string
+      if (!found && line.find("BO Stats Below") != std::string::npos) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportBOStats.cpp:78]: While this code is functionality, reading it can be a little bit difficult.  May I suggest the following alternative:
```
      if (!found) {
        if (line.find("BO Stats Below") != std::string::npos) 
          found = true;
	continue;
      }
```

File: src/runtime_src/core/tools/common/ReportBOStats.cpp (hunk 1 above)
Comment (rozumx) [reply src/runtime_src/core/tools/common/ReportBOStats.cpp:96]: Redundant data should NOT be placed into the JSON unless absolutely needed.  I don't believe that this is needed.   

If the TXT output needs additional computation needed that is perfectly acceptable.

File: src/runtime_src/core/tools/common/ReportBOStats.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        bo_pt.put("memory_used", mem_used);
+        bo_pt.put("memory_used_unit", bo_info[1].substr(bo_info[1].length()-2, 2));
+        bo_pt.put("memory_used_bytes", mem_used_bytes);
+      }
+      pt2_list.push_back(std::make_pair("", bo_pt));
+    }
+  } catch(const xrt_core::error& e) {
+    std::cerr << "ERROR: " << e.what() << std::endl;
+    std::cerr << "ERROR: bo_stats cmd - invalid format" << std::endl;
+    return; //Give error and let script continue
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportBOStats.cpp:114]: xbmgmt and xbutil returns a non-zero value whenever an error occurs.   This code masks this functionality.

Please throw an exception.

File: src/runtime_src/core/tools/common/ReportBOStats.cpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  } catch (...) {
+    std::cerr << "ERROR: bo_stats cmd - invalid format" << std::endl;
+    return; //Give error and let script continue
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportBOStats.cpp:117]: xbmgmt and xbutil returns a non-zero value whenever an error occurs.   This code masks this functionality.

Please throw an exception.

File: src/runtime_src/core/tools/xbutil2/SubCmdExamine.cpp [hunk 6]
```diff
@@ -55,28 +55,30 @@ namespace po = boost::program_options;
...
   static ReportCollection fullReportCollection = {
   // Common reports
     std::make_shared<ReportAie>(),
     std::make_shared<ReportAieShim>(),
-    std::make_shared<ReportMemory>(),
-    std::make_shared<ReportHost>(),
-    std::make_shared<ReportDynamicRegion>(),
-    std::make_shared<ReportDebugIpStatus>(),
     std::make_shared<ReportAsyncError>(),
+    std::make_shared<ReportBOStats>(),
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbutil2/SubCmdExamine.cpp:66]: Just wondering, why the change of order. Can't figure out the reason.

**Review by rozumx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/tools/common/ReportBOStats.cpp (hunk 1 above)
Comment (sarab96) [reply src/runtime_src/core/tools/common/ReportBOStats.cpp]: done

File: src/runtime_src/core/tools/common/ReportBOStats.cpp (hunk 3 above)
Comment (sarab96) [reply src/runtime_src/core/tools/common/ReportBOStats.cpp]: done

File: src/runtime_src/core/tools/common/ReportBOStats.cpp (hunk 4 above)
Comment (sarab96) [reply src/runtime_src/core/tools/common/ReportBOStats.cpp]: done

File: src/runtime_src/core/tools/common/ReportBOStats.cpp (hunk 5 above)
Comment (sarab96) [reply src/runtime_src/core/tools/common/ReportBOStats.cpp]: done

File: src/runtime_src/core/tools/xbutil2/SubCmdExamine.cpp (hunk 6 above)
Comment (sarab96) [reply src/runtime_src/core/tools/xbutil2/SubCmdExamine.cpp:66]: ascending order changes

**Review by rozumx — APPROVED** [review]
This looks great!  Thank you!

**Review by AShivangi — APPROVED** [review]
